src/orange/orangeinp/IntersectRegion.cc

```diff
@@ -334,6 +334,233 @@ void Ellipsoid::output(JsonPimpl* j) const
+{
+    insert_surface(Sense::outside, PlaneZ{-hh_});
+    insert_surface(Sense::inside, PlaneZ{hh_});
+
+    // Insert elliptical cylinder surface last, as a simple quadric with
+    // equation:
+    // r_y^2 x^2 + r_x^2 y^2 - r_x^2 r_y^2 = 0
+    real_type rx2 = ipow<2>(radii_[to_int(Axis::x)]);
+    real_type ry2 = ipow<2>(radii_[to_int(Axis::y)]);
+    real_type g = -rx2 * ry2;
+
+    Real3 abc{ry2, rx2, 0};
+    insert_surface(SimpleQuadric{abc, Real3{0, 0, 0}, g});
+
+    // Set exterior bbox
+    Real3 ex_halves{radii_[to_int(Axis::x)], radii_[to_int(Axis::y)], hh_};
+    insert_surface(Sense::inside, BBox{-ex_halves, ex_halves});
+
+    // Set an interior bbox (inscribed cuboid)
+    auto inv_sqrt_two = 1 / constants::sqrt_two;
+    Real3 in_halves{radii_[to_int(Axis::x)] * inv_sqrt_two,
+                    radii_[to_int(Axis::y)] * inv_sqrt_two,
+                    hh_};
+    insert_surface(Sense::outside, BBox{-in_halves, in_halves});
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write output to the given JSON object.
+ */
+void EllipticalCylinder::output(JsonPimpl* j) const
+{
+    to_json_pimpl(j, *this);
+}
+
+//---------------------------------------------------------------------------//
+// ELLIPTICAL CONE
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with lower/upper x- and y-radii and half-height in z.
+ *
+ * The lower radii are the x- and y-radii at the plane z = -hh. The
+ * upper radii are the x- and y-radii at the plane z = hh. There are several
+ * restrictions on these radii:
+ *
+ * 1) Either the lower or upper radii may be (0, 0); this is the only permitted
+ *    way for the elliptical cone to include the vertex.
+ * 2) The aspect ratio of the elliptical cross sections is constant. Thus, the
+ *    aspect ratio at z = -hh must equal the aspect ratio at z = hh.
+ * 3) Degenerate elliptical cones with lower_radii == upper_radii (i.e.,
+ *    elliptical cylinders) are not permitted.
+ * 4) Degenerate elliptical cones where lower or upper radii are equal to
+ *    (0, x) or (x, 0), where x is non-zero, are not permitted.
+ */
+EllipticalCone::EllipticalCone(Real2 const& lower_radii,
+                               Real2 const& upper_radii,
+                               real_type halfheight)
+    : lower_radii_{lower_radii}, upper_radii_{upper_radii}, hh_{halfheight}
+{
+    SoftZero soft_zero;
+    SoftEqual soft_equal;
+
+    auto X = to_int(Axis::x);
+    auto Y = to_int(Axis::y);
+
+    // True if either radius is negative
+    auto has_negative
+        = [&](Real2 const& radii) { return radii[X] < 0 || radii[Y] < 0; };
+
+    // True if radii is (0, 0)
+    auto is_vertex = [&](Real2 const& radii) {
+        return soft_zero(radii[X]) && soft_zero(radii[Y]);
+    };
+
+    // True if radii is (0, x) || (x, 0), where x != 0
+    auto is_partial_zero = [&](Real2 const& radii) {
```

Comment:
These lambdas don't need captures (if you use the built-in function `soft_zero` from SoftEqual.hh insted of the local functor.)

Reply:
They need captures for `X` and `Y`

Reply:
I think the local `soft_zero`, because it's using the default tolerance, is no different than the `celeritas::soft_zero` free function , which doesn't need to be bound.  Not super important, it's fine to capture blindly, but it's good knowledge to have.

Reply:
(done)

Reply:
Oops, looks like this breaks Windows. Shall I change it back?

```
D:\a\celeritas\celeritas\src\orange\orangeinp\IntersectRegion.cc(427): error C3493: 'X' cannot be implicitly captured because no default capture mode has been specified
```